internal/api/v1beta1/metaschema.go

```diff
@@ -0,0 +1,194 @@
+		}
+	}
+
+	metaschemaPB := transformMetaSchemaToPB(fetchedMetaSchema)
+
+	return &shieldv1beta1.GetMetaSchemaResponse{Metaschema: &metaschemaPB}, nil
+}
+
+func (h Handler) UpdateMetaSchema(ctx context.Context, request *shieldv1beta1.UpdateMetaSchemaRequest) (*shieldv1beta1.UpdateMetaSchemaResponse, error) {
+	logger := grpczap.Extract(ctx)
+
+	id := request.GetId()
+	if strings.TrimSpace(id) == "" {
+		return nil, grpcMetaSchemaNotFoundErr
+	}
+	if request.GetBody() == nil {
+		return nil, grpcBadBodyError
+	}
+
+	updateMetaSchema, err := h.metaSchemaService.Update(ctx, id, metaschema.MetaSchema{
+		Name:   request.GetBody().GetName(),
+		Schema: request.GetBody().GetSchema()})
+
+	if err != nil {
+		logger.Error(err.Error())
+		switch {
+		case errors.Is(err, metaschema.ErrInvalidDetail):
+			return nil, grpcBadBodyError
+		case errors.Is(err, metaschema.ErrInvalidID),
+			errors.Is(err, metaschema.ErrNotExist):
+			return nil, grpcMetaSchemaNotFoundErr
+		case errors.Is(err, metaschema.ErrConflict):
+			return nil, grpcConflictError
+		default:
+			return nil, grpcInternalServerError
+		}
+	}
+
+	MetaSchemaCache[updateMetaSchema.Name] = updateMetaSchema.Schema
+	metaschemaPB := transformMetaSchemaToPB(updateMetaSchema)
+	return &shieldv1beta1.UpdateMetaSchemaResponse{Metaschema: &metaschemaPB}, nil
+}
+
+func (h Handler) DeleteMetaSchema(ctx context.Context, request *shieldv1beta1.DeleteMetaSchemaRequest) (*shieldv1beta1.DeleteMetaSchemaResponse, error) {
+	logger := grpczap.Extract(ctx)
+
+	id := request.GetId()
+	if strings.TrimSpace(id) == "" {
+		return nil, grpcMetaSchemaNotFoundErr
+	}
+
+	name, err := h.metaSchemaService.Delete(ctx, id)
+	if err != nil {
+		logger.Error(err.Error())
+		switch {
+		case errors.Is(err, metaschema.ErrInvalidID),
+			errors.Is(err, metaschema.ErrNotExist):
+			return nil, grpcMetaSchemaNotFoundErr
+		default:
+			return nil, grpcInternalServerError
+		}
+	}
+
+	delete(MetaSchemaCache, name)
+	return &shieldv1beta1.DeleteMetaSchemaResponse{}, nil
+}
+
+func transformMetaSchemaToPB(from metaschema.MetaSchema) shieldv1beta1.MetaSchema {
+	return shieldv1beta1.MetaSchema{
+		Id:        from.ID,
+		Name:      from.Name,
+		Schema:    from.Schema,
+		CreatedAt: timestamppb.New(from.CreatedAt),
+		UpdatedAt: timestamppb.New(from.UpdatedAt),
+	}
+}
+
+// validates the metadata against the json-schema. In case metaschema doesn't exists in the cache, it will return nil (no validation)
+func validateMetadataSchema(mdata metadata.Metadata, schemaName string) error {
+	if MetaSchemaCache[schemaName] == "" {
+		return nil
+	}
+	metadataSchema := gojsonschema.NewStringLoader(MetaSchemaCache[schemaName])
+	providedSchema := gojsonschema.NewGoLoader(mdata)
+	results, err := gojsonschema.Validate(metadataSchema, providedSchema)
+	if err != nil {
+		return errors.Wrap(err, "failed to validate metadata")
+	}
+
+	if !results.Valid() {
+		return errors.New("metadata doesn't match the json-schema")
+	}
+	return nil
```

Comment:
We can move validate code to metaschema service code as well, expose a Validate method in it.